Lab9

nicko-lee opened the pull request:
Fixes #626 

Review of colleague's PR #617 

#### Changes proposed in this PR:

#### Implement lab9 requirements 
Implement ability to convert between Go in memory objects to JSON
representation and vice versa.

Implement ability to accept user input from CLI and read data from
a file.

Load data from json text file and store in an instance of storer.

Commit:
Implement lab9 requirements

Implement ability to convert between Go in memory objects to JSON
representation and vice versa.

Implement ability to accept user input from CLI and read data from
a file.

Load data from json text file and store in an instance of storer.

Commit:
Add tests to main_test.go

Commit:
Update README file to reflect changes in lab9

Commit:
Fix pesky scopelint error

Commit:
Fix pesky test coverage issue

Create new test with totally invalid json

Commit:
Fix final lint problem

Lint issue: should merge variable declaration with assignment on next line

nicko-lee commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,83 @@
+
+	// create some puppies loaded in from json file and store in store (no pun intended lol)
+	for _, pup := range puppies {
+		createdPuppy, _ := store.CreatePuppy(&pup)
```
Resolved

nicko-lee commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,87 @@
+
+func main() {
+	// var to store raw bytes read in from file
+	var fileData []byte
```
Resolved ya pesky bot lol

Commit:
Add 2 json files to read into program

kasunfdo reviewed (dismissed):
Looks good.
I would like to suggest including a json file that has a puppy with negative value.

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+
+func main() {
+	// parse flags and read json from file
+	fileData := parseFlagsAndLoadFile() // fileData var to store raw bytes read in from file
```
I would separate flag parsing and loading a file
```go
func main() {
  kingpin.Parse()
  filedata := loadFile(*filename)
}
```

A larger application might have multiple files it needs to load, or other weird things it needs to do with the parsed flags. Separating out like this makes it
1. Easier to test each part of the main function
2. Gives an error more specific to exactly what went wrong.

nicko-lee replied:
Have split this into 2 now. Good suggestion 🥇

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// parse flags and load file
+func parseFlagsAndLoadFile() (fileData []byte) {
```
Functions that can error out should return an error, not panic.

nicko-lee replied:
Could not figure out tests and test coverage and how to do this so for now have stuck with the panics. Might have a chat with you further about this.

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+// parse flags and load file
+func parseFlagsAndLoadFile() (fileData []byte) {
+	// parse cli flags and check for flag parse error
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
```
Return an error, let the main function decide whether to panic.

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+
+	// if data flag is supplied. Reading data from file and check for file argument error
+	f, err := os.Open(*fileName) // At this point Go knows about the file but not what's in it
+	if err != nil {
```
Return error

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	// reading what's in file. Raw as in it is a raw slice of bytes at this point
+	fileData, _ = ioutil.ReadAll(f) // no need to check file error as kingpin has checked it already
```
Kingpin only checks it if you use
`kingpin.Flag(stuff).File()`
You have
`kingpin.Flag("data", "").String()`

incidentally, using `.File()` removes the need to call `os.Open()`, since kingpin gives to an `**os.File`.

nicko-lee replied:
Cheers have changed from String() to ExistingFile()

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+	if err != nil {
+		printUsage()
+		panic(err)
+	}
```
`defer f.Close()` to make sure the file is closed after this function returns (you don't need this file open after this function returns)

nicko-lee replied:
Thanks man. Done. Although is this just for good practice? Cos in my case I don't think it matters if the file is still open?

anzboi replied:
Mainly good practice, should get in the habit of closing resources as soon as you are done with them.

anzboi commented on 09_json/nickolee/cmd/puppy-server/main_test.go:
```diff
@@ -0,0 +1,71 @@
+	var buf bytes.Buffer
+	out = &buf
+
+	args = []string{"-d", "./../../puppy-data/puppies.json"}
```
shouldn't need the `./` at the start of the path.

anzboi commented on 09_json/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// show usage guide if user is not providing args or providing the wrong args. Only applies to args not flags
+func printUsage() {
```
`kingpin.FatalUsage("format", args)` can replace this function.
Prints output of the form
```
main: error: My error errorstring
usage: main [<flags>]

Flags:
  --help       Show context-sensitive help (also try --help-long and --help-man).
  --file=FILE  
```

nicko-lee replied:
Thanks for the tip - tried doing this and removed my printUsage() func. It all worked well but i struggled to test it and hit the 100% coverage as well. Cos it did not panic so I couldn't test for assert.Panics. Nor did it return an error so I couldn't test for assert.Error. It simply PRINTED stuff out to stderr. So in the end I did away with this altogether...

anzboi commented on 09_json/nickolee/pkg/puppy/types.go:
```diff
@@ -0,0 +1,17 @@
+package puppy
+
+// Puppy data structure stores puppy properties
+type Puppy struct {
```
Nice 👍 
I would encourage you to experiment with `omitempty`. This will prevent fields from reaching output if they hold a zero value.
Whether any of these fields actually need it is up to you.

nicko-lee replied:
Thanks man will play around with it in my golang playground cheers !

anzboi commented on 09_json/nickolee/pkg/puppy/types_test.go:
```diff
@@ -0,0 +1,56 @@
+		expectedJSON string
+	}
+
+	tests := []test{
```
Great test cases.

nicko-lee replied:
Thanks 💯

anzboi commented on 09_json/nickolee/pkg/puppy/types_test.go:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	for _, test := range tests {
+		jsonRepresentation, _ := json.Marshal(test.puppyObject) // note that jsonReprentation is a slice of bytes
```
Would still recommend tabulated tests

```go
for name, test := range tests {
  t.Run(name, func(t *testing.T) {
    // test code
  })
}
```

anzboi replied:
I Tried

anzboi commented on 09_json/nickolee/puppy-data/invalid.json:
```diff
@@ -0,0 +1,27 @@
+Two households, both alike in dignity,
```
🤣

nicko-lee replied:
Haha 👍

anzboi reviewed (changes requested):
Great work so far
My main issue with the current code is main (excuse the pun). IMO auxillary functions used in main should return errors, and main should decide what to do with them.

I understand this makes it hard to hit 100% coverage, so if test coverage becomes an issue, the panics are ok.

Also, I think you should separate parsing the flags and reading the file should be separated.

Finally, you don't need `printUsage`, `kingpin.FatalUsage()` fills the same role, and updates as you change the flags.

Commit:
Fix PR feedback

1) add defer f.close() to close file
2) use kingpin.ExistingFile no kingpin.String
3) tried using kingpin.fatalusage() to replace my printusage() func
   but struggled to hit it for test coverage
4) Remove ./ at start of paths
5) separate flag parsing and loading a file

anzboi reviewed (approved):
LGTM

nicko-lee committed:
Merge branch 'master' into lab9
